### PR TITLE
Add QArchive 2.0.1

### PR DESCRIPTION
**Pull request by ericriff**
Specify library name and version:  **QArchive/2.0.1**

- [X] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [X] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [X] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [X] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/qarchive/all/conanfile.py
Comment (ericriff) [inline recipes/qarchive/all/conanfile.py:30]: This is ugly but easier to maintain than a patch file. I don't mind porting it to a patch file (once the changes are approved) if the community prefers it that approach.

**Comment by ericriff** [comment]
On my PC:
Maybe the hooks are outdated" :thinking: 
```
[HOOK - conan_hooks/hooks/conan-center.py] pre_build(): [FPIC MANAGEMENT (KB-H007)] OK. 'fPIC' option found and apparently well managed
[HOOK - conan_hooks/hooks/conan-center.py] pre_build(): [FPIC MANAGEMENT (KB-H007)] OK
```

**Comment by ericriff** [comment]
@uilianries 
On macOS I' getting
```
CMake Error at CMakeLists.txt:1 (cmake_minimum_required):
  CMake 3.17 or higher is required.  You are running version 3.16.4
```

Should I wait for the docker image of macOS to be updated or pull `cmake` as a `build_requirement` ?

**Comment by ericriff** [comment]
On Windows builds I'm getting 
`_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in test_package.obj`

and 

`error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MTd_StaticDebug' in test_package.obj`

And it looks like somehow debug and release stuff are being mixed? 
Can I get some help from a Windows guru? :pray:

File: recipes/qarchive/all/conanfile.py
Comment (prince-chrismc) [inline recipes/qarchive/all/conanfile.py:31]: maybe use a build_requires?

Comment (ericriff) [reply recipes/qarchive/all/conanfile.py:31]: I'm fine with both approaches. I can change it to that once I get a build to pass on CCI. Windows is being annoying.

**Comment by ericriff** [comment]
This is as far as I can push the windows builds, I'm not sure what is going on.
Any help from a Qt/Windows guru will be greatly appreciated :)

**Comment by ericLemanissier** [comment]
AFAIU, it fails because of an upstream bug: [the DiskExtractor class](https://github.com/antony-jr/QArchive/blob/master/include/qarchivediskextractor.hpp#L12) (and probably others) does not use the import/export incantations needed on windows: https://doc.qt.io/qt-5/sharedlibrary.html#using-symbols-from-shared-libraries

**Comment by prince-chrismc** [comment]
Disable windows shared?

**Comment by ericriff** [comment]
I've disable the shared builds on linux and I'll report the error to the dev upstream.

**Comment by ericriff** [comment]
This is ready to review now.

**Comment by ericriff** [comment]
@jgsogo Do you know why CCI doesn't like me anymore? It won't build my PR regardless of how many times I close/reopen it. :cry:

**Comment by antony-jr** [comment]
> @jgsogo @uilianries Do you know why CCI doesn't like me anymore? It won't build my PR regardless of how many times I close/reopen it. cry

@ericriff Try bumping to version 2.0.4, that version has only typo fixes but more properly uses the import and export that is required in WIndows builds. v2.0.3 might also just work fine.

**Comment by jgsogo** [comment]
Yesterday we faced an issue in CI infrastructure. The symptoms were all jobs failing because of _"no space left on device"_, but so far we have plenty of space 🤷‍♂️ . We know that restarting the cluster fixes the issue, but as it was during working hours of the infrastructure team we asked them to investigate further... and they finally found that it was not disk size, but inodes. Now we need to design a solution to prevent this error from happening again.

I'm really sorry for the inconvenience, but it was important to find the cause of the error (it isn't the first time it happens).

**Comment by antony-jr** [comment]
> Yesterday we faced an issue in CI infrastructure. The symptoms were all jobs failing because of _"no space left on device"_, but so far we have plenty of space man_shrugging . We know that restarting the cluster fixes the issue, but as it was during working hours of the infrastructure team we asked them to investigate further... and they finally found that it was not disk size, but inodes. Now we need to design a solution to prevent this error from happening again.
> 
> I'm really sorry for the inconvenience, but it was important to find the cause of the error (it isn't the first time it happens).

It's very hard to run your own CI in the first place :smile:.

**Comment by antony-jr** [comment]
@ericriff Try the next version (2.0.4), I think the next version might fix the above error with non shared builds. It seems 2.0.3 succeeded with the shared builds.

**Comment by ericriff** [comment]
@antony-jr it looks like windows static is failing now :thinking:

**Comment by antony-jr** [comment]
> @antony-jr it looks like windows static is failing now thinking

I just found the fix. Windows is definitely ugly. So we need to avoid import/export when using static libraries but these has to somehow controlled within the header without any definition. Quazip seems to depend on the declaration of **definition** if it is a static build but that is cumbersome and ugly. Let me see if I could do some CMake magic to automatically handle this definitions.

**Comment by antony-jr** [comment]
@ericriff Please try v2.0.5, it should work for sure this time(I hope).

**Comment by antony-jr** [comment]
@ericriff In the test package, you try to extract ```Test.7z``` which does not exists so error signal must have emitted but there is no handlers for error signal and only waits for the finished signal so **the application never exits and the given time exceeds and the test fails?** But I don't know how it passed in other builds, why it did not complain?? Maybe the cancel got to the extractor faster than start.

You can use the unit tests if you want. Or download test 7z archive and place it next the test application when executing.
Or you have to use the compressor to create a simple archive and extract it and delete it when finished.

```
 $ cmake -DBUILD_TESTS=ON . 
 $ cd tests 
 $ ./QArchiveTests
```

**Comment by antony-jr** [comment]
@ericriff How about a test package like this(This does not require any test files, it creates it automatically),

```
#include <QCoreApplication>
#include <QTemporaryDir>
#include <QFile>
#include <QDir>
#include <QDebug>
#include <QArchive>


int main(int ac, char **av) {
        QCoreApplication app(ac, av);
        QTemporaryDir dir;
        QString data = QString::fromUtf8("Some Text to Compress.");
        QFile textFile;
        textFile.setFileName(dir.path() + QDir::separator() + "TextFile.txt");
        if(!textFile.open(QIODevice::WriteOnly)) {
                return -1;
        }

        textFile.write(data.toUtf8());
        textFile.close();

        QArchive::DiskCompressor compressor;
        QArchive::DiskExtractor extractor;

        compressor.setFileName(dir.path() + QDir::separator() + "Test.zip");
        compressor.addFiles(QString::fromUtf8("Output.txt") , dir.path() + QDir::separator() + "TextFile.txt");

        extractor.setArchive(dir.path() + QDir::separator() + "Test.zip");
        extractor.setOutputDirectory(dir.path());

        QObject::connect(&compressor, &QArchive::DiskCompressor::finished,
        [&]() {
                extractor.start();
        });

        QObject::connect(&extractor, &QArchive::DiskExtractor::finished,
        [&]() {
                /// Check if the data is correct.
                QFile file;
                file.setFileName(dir.path() + QDir::separator() + "Output.txt");
                if(!file.open(QIODevice::ReadOnly)){
                        qDebug() << "Cannot Open Output.txt!";
                        app.quit();
                }

                auto got = QString(file.readAll());
                if(got != data) {
                        qDebug() << "Data does not Match: " << got;
                        app.quit();
                }
                qDebug() << "Data: " << got;
                qDebug() << "Test Finished.";
                app.quit();
        });

        compressor.start();
        return app.exec();
}
```

**Comment by ericriff** [comment]
> @ericriff In the test package, you try to extract `Test.7z` which does not exists so error signal must have emitted but there is no handlers for error signal and only waits for the finished signal so **the application never exits and the given time exceeds and the test fails?** But I don't know how it passed in other builds, why it did not complain??
> 
> You can use the unit tests if you want. Or download test 7z archive and place it in the working directory when executing.
> Or you have to use the compressor to create a simple archive and extract it and delete it when finished.
> 
> ```
>  $ cmake -DBUILD_TESTS=ON . 
>  $ cd tests 
>  $ ./QArchiveTests
> ```

If you take a closer look at the test package I removed the last line (compared with your example).
I replaced `return app.exec();` with `return 0`. My expectation was for that change to prevent the actual decompression from running.
https://github.com/conan-io/conan-center-index/pull/4774/files#diff-75f143c74f991b79547834e5edf2fe32534fbbaecf5a381bfaf760f8b7d156f5R19

Usually test packages in conan are used to test that you can actually include the public headers and link to the lib correctly, they don't always test the actual functionality of the library. So I guessed that if I removed that line I could get away without including an actual 7z file.

Regardless I tried the test you proposed and, for linux, it works when I build static but with shared nothing happens and it get stuck (not even the first log line gets printed).

Using the current code CCI builds show error 3221225781which google says it is a missing dll error :thinking:

**Comment by ericriff** [comment]
I'vemanually built it on a Windows PC, shared, and I get
aqarchive
So the missing dll is the actual QAarchive lib. Will look into it.

**Comment by ericriff** [comment]
I've modified the CMakelists on the test package to use FindQArchive, now the log shows
```
-- Conan: Using autogenerated FindQArchive.cmake
-- Found QArchive: 2.0.5 (found version "2.0.5")
-- Library QArchive found C:/Users/ericr/.conan/data/qarchive/2.0.5/asd/asd/package/665aa1bc6974aa7931843b6ba4d8715d732d7d58/lib/QArchive.lib
-- Found: C:/Users/ericr/.conan/data/qarchive/2.0.5/asd/asd/package/665aa1bc6974aa7931843b6ba4d8715d732d7d58/lib/QArchive.lib
```
But I get the same error. It is weird that we're building the shared version of the library but conan tries to use the static version instead.
On package/lib folder there is a QArchive.dll and QArchive.lib.

I'm not sure what is going on.

**Comment by antony-jr** [comment]
> I've manually built it on a Windows PC, shared, and I get
> aqarchive
> So the missing dll is the actual QAarchive lib. Will look into it.

Maybe copy the QArchive.dll to the test_package.exe? I have no clue.

**Comment by prince-chrismc** [comment]
To my knowledge, it needs to be in the `packag_folder/bin` for it to work. Sometimes the upstream install sticks it in the wrong folder.

You might was to check the build logs to see if the `package()` reports a DLL.

**Comment by ericriff** [comment]
> To my knowledge, it needs to be in the `packag_folder/bin` for it to work. Sometimes the upstream install sticks it in the wrong folder.
> 
> You might was to check the build logs to see if the `package()` reports a DLL.

The build logs does show the dll being installed, and with no funky names or anything.
```

**Comment by prince-chrismc** [comment]
> If I manually move the `dll` to test_package/bin` it works. But isn't that something that conan should handle?

Sadly Conan can not correct bugs in upstream projects. Could it be smart enough to automatically adjust for it... maybe but I also appreciate it forcing the correct behavoir

**Comment by ericriff** [comment]
> > If I manually move the `dll` to test_package/bin` it works. But isn't that something that conan should handle?
> 
> Sadly Conan can not correct bugs in upstream projects. Could it be smart enough to automatically adjust for it... maybe but I also appreciate it forcing the correct behavior
> 
> EDIT:
> 
> Conan automatically includes the `bin` folder of the package, you just need to move that `dll` to the correct location

Interesting.
`package/hash/lib` isn't the right folder for dlls? I'm sorry if it is a dumb question, I'm not a Windows developer.

**Comment by prince-chrismc** [comment]
> `package/hash/lib` isn't the right folder for dlls

That's correct!

I am shamefully a Windows developer... I do it for the money 🙊

**Comment by ericriff** [comment]
> > `package/hash/lib` isn't the right folder for dlls
> 
> That's correct!
> 
> I am shamefully a Windows developer... I do it for the money

Good to know that you're a windows dev, now I know who to ask when CCI blows up with windows build haha
Where should the dlls be packaged then?

**Comment by prince-chrismc** [comment]
`bin` folder!

Always a pleasure to help 😛

**Comment by ericriff** [comment]
> `bin` folder!
> 
> Always a pleasure to help

Thanks!
@antony-jr could you fix this upstream?
Apparently Windows wants the `dll` to be installed into the `bin` folder.

**Comment by prince-chrismc** [comment]
https://github.com/conan-io/conan-center-index/blob/689e2663fe41689259f61a5dbd3fe245e24c03b9/recipes/libftdi/0.x/patches/0001-cmake-fixes.patch#L49

**Comment by antony-jr** [comment]
> > `bin` folder!
> > Always a pleasure to help
> 
> Thanks!
> @antony-jr could you fix this upstream?
> Apparently Windows wants the `dll` to be installed into the `bin` folder.

Only Windows depends on this? All other build seems to work fine. Should we need this change or can we use a patch like above?

**Comment by antony-jr** [comment]
@ericriff I think you can specify CMake for the lib directory, Can you try in your local build?

```
 $ cmake -DCMAKE_INSTALL_LIBDIR=bin ..
```

**Comment by ericriff** [comment]
@antony-jr I've manually patched the CMakeLists here https://github.com/conan-io/conan-center-index/pull/4774/commits/fac3d7230fe3066e3055c9c4bb57c4e5dfb24430
And it worked for windows, locally. Lets see if CCI is able to build for the rest of the platforms.
If it passes it would be ideal for me to revert that commit so you can adopt those changes upstream.

**Comment by ericriff** [comment]
@antony-jr All the builds passed! :D
Thank you so much @prince-chrismc for pointing out the issue!

File: recipes/qarchive/all/conanfile.py
Comment (prince-chrismc) [inline recipes/qarchive/all/conanfile.py:29]: You can explicitly _always_ set this in the cmake_wrapper since it handles all the logic about applying it

**Review by prince-chrismc — CHANGES_REQUESTED** [review]
You are missing this snippet too https://github.com/conan-io/conan-center-index/blob/689e2663fe41689259f61a5dbd3fe245e24c03b9/recipes/http_parser/all/conanfile.py#L34-L36

**Comment by ericriff** [comment]
>  def configure(self): 
>      if self.options.shared: 
>          del self.options.fPIC

It is there somewhere, otherwise the CCI would complain about it.

**Comment by antony-jr** [comment]
@ericriff So I'm thinking about doing the requested changes and create a new final release v2.0.1 with all the changes up till now and delete every other release as it is very redundant. I will ping you when this is done.

**Comment by antony-jr** [comment]
@ericriff I've purged all redundant versions and now only **v2.0.1** exists with all the required changes that requires for integration with conan.

**Comment by ericriff** [comment]
@antony-jr Thanks! There are some issues with CCI now, it is failing to build because of `apt` errors. I'll update the branch later today.
Thanks for all the help!

**Comment by prince-chrismc** [comment]
My comment above still applies... They are ever so slightly different 🔍 

image

**Comment by ericriff** [comment]
@prince-chrismc done :)

**Review by prince-chrismc — APPROVED** [review]
I think you can move the patches to patches =)

But LGTM!

**Comment by prince-chrismc** [comment]
Just revert to the tools replace.... There's a ton of CMake version and getting ALL to recompile will be a pain

**Comment by ericriff** [comment]
Are you saying my PR to bump the SSL version won't be smooth?

**Comment by prince-chrismc** [comment]
Absolutely it will take time. I would be surprised if it passed.

**Review by prince-chrismc — APPROVED** [review]

**Review by mathbunnyru — APPROVED** [review]

**Review by SSE4 — APPROVED** [review]

**Comment by antony-jr** [comment]
Thanks for your efforts @ericriff, if you want you can open a PR at QArchive to include instructions to use this conan package in the documentation or README. Thanks.

**Pull request merged**
